src/box/lua/config/applier/roles.lua

```diff
@@ -1,16 +1,78 @@
 local log = require('internal.config.utils.log')
+local fiber = require('fiber')
 
-local last_loaded = {}
-local last_roles_ordered = {}
+local roles_state = {
+    -- Loaded roles, where the key is the role name and the value is the
+    -- role object.
+    last_loaded = {},
+
+    -- Ordered list of roles, where the order is determined by the
+    -- dependencies between roles.
+    last_roles_ordered = {},
+
+    -- The box.status watcher is used to call on_box_status callbacks.
+    box_status_watcher = nil,
+
+    -- Last values received from box.status watcher.
+    last_box_status_value = nil,
+
+    -- Last applied config.
+    last_config = nil,
+}
+
+-- Call all the 'on_box_status' callbacks from roles in order.
+local function call_on_box_status_callbacks(config, value)
+    for _, role_name in ipairs(roles_state.last_roles_ordered) do
+        local role = roles_state.last_loaded[role_name]
+        if role.on_box_status ~= nil then
+            log.verbose(('roles.on_box_status: calling callback for role ' ..
+                        '"%s"'):format(role_name))
+            local ok, err = pcall(role.on_box_status,
+                                  config, 'box.status', value)
+            if not ok then
+                log.error(('roles.on_box_status: callback for role ' ..
+                            '"%s" failed: %s'):format(role_name, err))
+            end
+        end
+    end
+end
+
+local function box_status_watcher()
+    return box.watch('box.status', function(key, value)
+        assert(key == 'box.status')
+
+        -- Store the last value to be able to call on_box_status callbacks
+        -- when the config is changed.
+        roles_state.last_box_status_value = table.deepcopy(value)
+
+        -- Config is not fully applied yet, don't call on_box_status
+        -- callbacks. We will call them later in post_apply.
+        if roles_state.last_config == nil then
+            return
+        end
+
+        call_on_box_status_callbacks(roles_state.last_config, value)
+    end)
+end
 
 local function apply(_config)
-    log.verbose('roles.apply: do nothing')
+    if roles_state.box_status_watcher == nil then
+        roles_state.box_status_watcher = box_status_watcher()
+
+        local deadline = fiber.time() + 10
+        while roles_state.last_box_status_value == nil do
+            fiber.sleep(0.01)
+            if fiber.time() > deadline then
+                error('Timeout reached while waiting for box_status_watcher')
```

Comment:
Nit: I report all the error from config's code using `error(<...>, 0`, because this way a user don't see some internal files/lines in the report.

If it is a user side error, for example, from config validation. Here I don't mind, just highlighted my personal preference. Feel free to ignore.

Reply:
Fixed!